pkg/notification/notice.go

```diff
@@ -0,0 +1,170 @@
+	currentVersion   string
+	latestVersion    updateResponse
+}
+
+// NewVersionChecker creates a new VersionChecker with the default
+// updates API URL. The URL can be overridden by passing an Option
+// to the NewVersionChecker function.
+func NewVersionChecker(opts ...Option) *VersionChecker {
+	v := &VersionChecker{
+		updatesApi:     "https://check.trivy.cloud/updates",
+		currentVersion: app.Version(),
+	}
+
+	for _, opt := range opts {
+		opt(v)
+	}
+	return v
+}
+
+// RunUpdateCheck makes a best efforts request to determine the
+// latest version and any announcements
+// Logic:
+// 1. if skipUpdateCheck is true AND telemetryDisabled are both true, skip the request
+// 2. if skipUpdateCheck is true AND telemetryDisabled is false, run check with metric details but suppress output
+// 3. if skipUpdateCheck is false AND telemetryDisabled is true, run update check but don't send any metric identifiers
+func (v *VersionChecker) RunUpdateCheck(ctx context.Context, args []string) {
+	logger := log.WithPrefix("notification")
+
+	if v.skipUpdateCheck && v.telemetryDisabled {
+		logger.Debug("Skipping update check and metric ping")
+		return
+	}
+
+	go func() {
+		args = getFlags(args)
+		client := &http.Client{
+			Timeout: 3 * time.Second,
+		}
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, v.updatesApi, http.NoBody)
+		if err != nil {
+			logger.Warn("Failed to create a request for Trivy api", log.Err(err))
+			return
+		}
+
+		// if the user hasn't disabled metrics, send the anonymous information as headers
+		if !v.telemetryDisabled {
+			req.Header.Set("Trivy-Identifier", uniqueIdentifier())
+			req.Header.Set("Trivy-Command", strings.Join(args, " "))
+			req.Header.Set("Trivy-OS", runtime.GOOS)
+			req.Header.Set("Trivy-Arch", runtime.GOARCH)
+		}
+
+		req.Header.Set("User-Agent", fmt.Sprintf("trivy/%s", v.currentVersion))
+		resp, err := client.Do(req)
+		if err != nil || resp.StatusCode != http.StatusOK {
+			logger.Debug("Failed getting response from Trivy api", log.Err(err))
+			return
+		}
+
+		defer resp.Body.Close()
+		if err := json.NewDecoder(resp.Body).Decode(&v.latestVersion); err != nil {
+			logger.Debug("Failed to decode the Trivy response", log.Err(err))
+			return
+		}
+
+		// enable priting if update allowed and quiet mode is not set
+		if !v.skipUpdateCheck && !v.quiet {
+			v.responseReceived = true
+		}
+		v.done = true
+	}()
+}
+
+// PrintNotices prints any announcements or warnings
+// to the output writer, most likely stderr
+func (v *VersionChecker) PrintNotices(output io.Writer) {
+	if !v.responseReceived {
+		return
+	}
+
+	var notices []string
+
+	notices = append(notices, v.latestVersion.Warnings...)
+	for _, announcement := range v.latestVersion.Announcements {
+		if time.Now().Before(announcement.ToDate) && time.Now().After(announcement.FromDate) {
+			notices = append(notices, announcement.Announcement)
+		}
+	}
+
+	if v.currentVersion != v.latestVersion.Trivy.LatestVersion {
```

Comment:
Can we use [LatestVersion()](https://github.com/aquasecurity/trivy/blob/e58ccddd47cdc398f905583224c2ab8a2a07443c/pkg/notification/notice.go#L147) function here?

Reply:
same for Announcements()

Reply:
done

Reply:
I hadn't :-) I was still thinking about the short vs long flags. Done now